grpc/rust/compile/tonic_compile.bzl

```
@@ -0,0 +1,70 @@
+#
+# Copyright (C) 2022 Vaticle
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+
+def _rust_tonic_compile_impl(ctx):
+    inputs = ctx.attr.protoc.files.to_list()
+    proto_paths_str = ""
+    for src in ctx.attr.srcs:
+        inputs.append(src[ProtoInfo].direct_sources[0])
+        proto_paths_str = proto_paths_str + src[ProtoInfo].direct_sources[0].path + ";"
```

This could be simplified to
```suggestion
    sources = [src[ProtoInfo].direct_sources[0] for src in ctx.attr.srcs]
    inputs = ctx.attr.protoc.files.to_list() + sources
    proto_paths_str = ";".join([src.path for src in sources])
```
The last line could be inlined in the `ctx.actions.run()` call as well.

---

Yeah, good point. To be honest I couldn't remember if Starlark supported list comprehensions so I lazily just went with loops. Updated to use comprehensions.